src/coreComponents/constitutive/fluid/BlackOilFluid.cpp

```diff
@@ -31,76 +26,443 @@ using namespace dataRepository;
+                                             Bousat[branchSize - 1],
+                                             m_PVTO.m_maxRelativePressure );
+      real64 const visc = linearExtrapolation( Pusat[branchSize - 2],
+                                               viscusat[branchSize - 2],
+                                               Pusat[branchSize - 1],
+                                               viscusat[branchSize - 1],
+                                               m_PVTO.m_maxRelativePressure );
+      m_PVTO.m_undersaturatedBo[i].emplace_back( Bo );
+      m_PVTO.m_undersaturatedViscosity[i].emplace_back( visc );
+      m_PVTO.m_undersaturatedPressure[i].emplace_back( m_PVTO.m_maxRelativePressure );
+    }
+  }
 }
 
-void BlackOilFluid::postProcessInput()
+void BlackOilFluid::refineTableAndCopy( localIndex const nLevels )
 {
-  // TODO maybe use different names?
-  m_componentNames = m_phaseNames;
+  std::vector< real64 > refineP; // TODO:change to array1d
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
+    for( localIndex j = 0; j < m_PVTO.m_undersaturatedPressure[i].size(); ++j )
+    {
+      refineP.emplace_back( m_PVTO.m_undersaturatedPressure[i][j] );
+    }
+  }
+  std::sort( refineP.begin(), refineP.end() );
+  refineP.erase( std::unique( refineP.begin(), refineP.end() ), refineP.end() );
+
+  // linspace
+  real64 const maxVal = *std::max_element( refineP.begin(), refineP.end() );
+  array1d< real64 > refinePP( refineP.size() );
+  for( localIndex i = 0; i < localIndex( refineP.size() ); ++i )
+  {
+    refinePP[i] = refineP[i];
+  }
+
+  refinePP.resize( nLevels );
+  for( localIndex i = 0; i != nLevels; ++i )
+  {
+    refinePP[i] = i * ( maxVal / ( nLevels - 1 ) );
+  }
 
-  MultiFluidPVTPackageWrapper::postProcessInput();
+  localIndex const ppSize = refinePP.size();
+  array1d< real64 > propRefined( ppSize );
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
 
-  localIndex const NP = numFluidPhases();
+    // Bo
+    interpolation1( m_PVTO.m_undersaturatedPressure[i], m_PVTO.m_undersaturatedBo[i], refinePP, propRefined );
+    m_PVTO.m_undersaturatedBo[i].resize( ppSize );
+    for( localIndex j = 0; j < ppSize; ++j )
+    {
+      m_PVTO.m_undersaturatedBo[i][j] = propRefined[j];
+    }
 
-  checkInputSize( m_surfaceDensities, NP, viewKeyStruct::surfaceDensitiesString() );
-  checkInputSize( m_tableFiles, NP, viewKeyStruct::tableFilesString() );
+    // viscosity
+    interpolation1( m_PVTO.m_undersaturatedPressure[i], m_PVTO.m_undersaturatedViscosity[i], refinePP, propRefined );
+    m_PVTO.m_undersaturatedViscosity[i].resize( ppSize );
+    for( localIndex j = 0; j < ppSize; ++j )
+    {
+      m_PVTO.m_undersaturatedViscosity[i][j] = propRefined[j];
+    }
+
+    for( localIndex j = 0; j < ppSize; ++j )
+    {
+      m_PVTO.m_undersaturatedPressure[i].resize( ppSize );
+      m_PVTO.m_undersaturatedPressure[i][j] = refinePP[j];
```

Comment:
Is this resizing `m_PVTO.m_undersaturatedPressure[i]` on inner each loop iteration?